src/contracts/middleware/BLSPubkeyRegistry.sol

```diff
@@ -0,0 +1,208 @@
+        // emit event so offchain actors can update their state
+        emit PubkeyAdded(operator, pubkey);
+        return pubkeyHash;
+    }
+
+    /**
+     * @notice Deregisters the `operator`'s pubkey for the specified `quorumNumbers`.
+     * @param operator The address of the operator to deregister.
+     * @param quorumNumbers The quourm numbers the operator is deregistering from, where each byte is an 8 bit integer quorumNumber.
+     * @param pubkey The public key of the operator.
+     * @dev access restricted to the RegistryCoordinator
+     * @dev Preconditions:
+     *         1) `quorumNumbers` has no duplicates
+     *         2) `quorumNumbers.length` != 0
+     *         3) `quorumNumbers` is ordered in ascending order
+     *         4) the operator is not already deregistered
+     *         5) `quorumNumbers` is the same as the parameter use when registering
+     *         6) `pubkey` is the same as the parameter used when registering
+     */   
+    function deregisterOperator(address operator, bytes memory quorumNumbers, BN254.G1Point memory pubkey) external onlyRegistryCoordinator returns(bytes32){
+        bytes32 pubkeyHash = BN254.hashG1Point(pubkey);
+
+        // update each quorum's aggregate pubkey
+        _processQuorumApkUpdate(quorumNumbers, pubkey.negate());
+        // update the global aggregate pubkey
+        _processGlobalApkUpdate(pubkey.negate());
+        // emit event so offchain actors can update their state
+        emit PubkeyRemoved(operator, pubkey);
+        return pubkeyHash;
+    }
+
+    /// @notice Returns the current APK for the provided `quorumNumber `
+    function getApkForQuorum(uint8 quorumNumber) external view returns(BN254.G1Point memory) {
+        return quorumApk[quorumNumber];
+    }
+
+    /// @notice Returns the `ApkUpdate` struct at `index` in the list of APK updates for the `quorumNumber`
+    function getApkUpdateForQuorumByIndex(uint8 quorumNumber, uint256 index) external view returns (ApkUpdate memory){
+        return quorumApkUpdates[quorumNumber][index];
+    }
+
+    /**
+     * @notice get hash of the apk of `quorumNumber` at `blockNumber` using the provided `index`;
+     * called by checkSignatures in BLSSignatureChecker.sol.
+     * @param quorumNumber is the quorum whose ApkHash is being retrieved
+     * @param blockNumber is the number of the block for which the latest ApkHash muust be retrieved
+     * @param index is the provided witness of the onchain index calculated offchain
+     */
+    function getApkHashForQuorumAtBlockNumberFromIndex(uint8 quorumNumber, uint32 blockNumber, uint256 index) external returns (bytes32){
+        ApkUpdate memory quorumApkUpdate = quorumApkUpdates[quorumNumber][index];
+        _validateApkHashForQuorumAtBlockNumber(quorumApkUpdate, blockNumber);
+        return quorumApkUpdate.apkHash;
+    }
+
+	/**
+     * @notice get hash of the apk among all quourums at `blockNumber` using the provided `index`;
```

Comment:
typo again: 'quourums' => 'quorums'
I would also recommend changing the wording to 'the global apk' to be consistent, although I think we should also define the 'global apk' more clearly, perhaps in the top-level contract description?